app-tracking-protection/vpn-store/src/main/java/com/duckduckgo/mobile/android/vpn/store/VpnDatabaseCallback.kt

```diff
@@ -24,32 +24,50 @@ import com.duckduckgo.common.utils.DispatcherProvider
 import com.duckduckgo.mobile.android.vpn.trackers.*
 import com.squareup.moshi.JsonAdapter
 import com.squareup.moshi.Moshi
-import java.util.*
 import javax.inject.Provider
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.ExperimentalCoroutinesApi
-import kotlinx.coroutines.asExecutor
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import logcat.logcat
 
 @OptIn(ExperimentalCoroutinesApi::class)
 internal class VpnDatabaseCallback(
     private val context: Context,
     private val vpnDatabase: Provider<VpnDatabase>,
     private val dispatcherProvider: DispatcherProvider,
+    private val coroutineScope: CoroutineScope,
+    private val mutex: Mutex,
 ) : RoomDatabase.Callback() {
 
     override fun onCreate(db: SupportSQLiteDatabase) {
         super.onCreate(db)
-        ioThread {
-            prepopulateAppTrackerBlockingList()
-            prepopulateAppTrackerExclusionList()
-            prepopulateAppTrackerExceptionRules()
+        coroutineScope.launch(dispatcherProvider.io()) {
+            mutex.withLock {
+                if (vpnDatabase.get().vpnAppTrackerBlockingDao().getExclusionListMetadata()?.eTag == null) {
+                    logcat { "VPN db onCreate: pre-populating db" }
+                    // only pre-populate when there's no blocklist
+                    prepopulateAppTrackerBlockingList()
+                    prepopulateAppTrackerExclusionList()
+                    prepopulateAppTrackerExceptionRules()
+                } else {
+                    logcat { "VPN db onCreate: SKIP pre-populating db" }
+                }
+            }
         }
     }
 
     override fun onDestructiveMigration(db: SupportSQLiteDatabase) {
-        ioThread {
-            prepopulateAppTrackerBlockingList()
-            prepopulateAppTrackerExclusionList()
-            prepopulateAppTrackerExceptionRules()
+        coroutineScope.launch(dispatcherProvider.io()) {
+            mutex.withLock {
+                if (vpnDatabase.get().vpnAppTrackerBlockingDao().getExclusionListMetadata()?.eTag == null) {
```

Comment:
I think we want the `getTrackerBlocklistMetadata` here and maybe as an extra safety mechanism sanity-check that there are actually some entities in. Might be worth adding a test case for this branch as well?
```suggestion
                if (vpnDatabase.get().vpnAppTrackerBlockingDao().getTrackerBlocklistMetadata()?.eTag == null &&
                    vpnDatabase.get().vpnAppTrackerBlockingDao().hasTrackerEntities()
                ) {
```

Reply:
ah good catch, this was a bug indeed